### PR TITLE
NioLoop

### PR DESCRIPTION
**Pull request by sagenschneider**
Initial work for #18    See NioLoopTest  for example of this PR working.

Note: this does break existing tests.   However, I'd like to discuss architecture of the driver before doing too much more, as:
  - should provide separation between the asynchronous API "promises" of the ADBA spec and the actual submissions going over the wire
  - current state mechanism is considering only a request at a time, however as per NioLoopTest is shows that potential for pipelining requests (allowing multiple requests to be in flight to improve performance).   Unashamedly, this is why I've taking interest in this project, so will need some convincing why pipelining should not be supported ;)

Note: I'm not fully across the ADBA spec.  I currently write high performance networking code (HTTP driver is currently fastest for sending XML responses in TechEmpower - due to very good buffer management and low CPU costs to parse HTTP messages).   I want to help in this regards to get the pgsql2 driver high performing :)

I'll write my thoughts shortly on #18 on what I'd see the architecture might be to help get discussions going :)  

Note: again raising this PR for you to have some level of working code.  It's only what I could put together in limited time over the last two nights.  I, however, hope it helps you understand what I'm thinking and I'll put up a more complete description on #18

### CODE REVIEW

**Comment by sagenschneider** [comment]
@alexanderkjall   also, I'm getting some formatting differences.  I've changed to use 2 spaces.   I'm using Eclipse Photon for development.  Might you have a format file that I could use so we don't get format differences clouding actual code changes?

File: src/main/java/org/postgresql/sql2/execution/NioService.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
+/*
```
[the remark this answers is not in the document]
Comment (sagenschneider) [reply src/main/java/org/postgresql/sql2/execution/NioService.java]: Yep.... my bad...   my IDE automatically writes things in for OfficeFloor (http://officefloor.net).    Happy to work to licensing for pgsql2.   I'll get it removed :+1:

**Comment by sagenschneider** [comment]
Ok, removed the invalid licensing.  Thanks @cretz .   Certainly not my intention, just over sight by me with my IDE settings.

File: src/main/java/org/postgresql/sql2/PGConnectionProperties.java [hunk 2]
```diff
@@ -360,7 +362,12 @@
   /**
    *
    */
-  TIMEZONE(String.class, "", false);
+  TIMEZONE(String.class, "", false),
+
+  /**
+   * Allows specifying the {@link NioLoop}.
+   */
+  NIO_LOOP(String.class, "", false);
```
Comment (alexanderkjall) [inline src/main/java/org/postgresql/sql2/PGConnectionProperties.java]: This should be 
NIO_LOOP(NioLoop.class, null, false);

**Comment by alexanderkjall** [comment]
Hi, first some administrative things:

Formatting: The pgjdbc project uses the google java formatting style, and I have been using a lazy variant of it. If you don't dislike it I'll add a checkstyle step to gradle ( with these rules https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml ).

Tests: I have been using the testcontainers project in order to be able to have very precise control over the postgresql server that the tests run against. This becomes specially important with the testing for tls connections (that I'm currently working on) and to be able to run tests on many different versions of the server easily. But I'm not 100% happy with them, it seems like the connection between the test framework and docker sometimes collapses, so I haven't configured any external testing with for example travis, in order to not get flooded with false negatives.

And now your code:

Regarding NioLoop: I think it's a great idea to abstract the loop into it's own object so that it can be reused. 

I noticed that you did a new Thread(...).start() and if I recall correctly that is not allowed in the J2EE spec, I think that needs to be replaced with some sort of ExecutorService.

Should the loop maybe be exposed through a get method, so that you don't have to inject it in order to reuse it?

Pipelining: We should really support this, but it won't always be used as a request can have a future produced by an old request as a parameter.

File: src/main/java/org/postgresql/sql2/PGConnectionProperties.java (hunk 2 above)
Comment (sagenschneider) [reply src/main/java/org/postgresql/sql2/PGConnectionProperties.java]: Yes, happy to change.   Focus for me was to get something together to prove the concepts.  Will fix :)

**Comment by sagenschneider** [comment]
Yep, as long as I have formatting rules in my IDE (Eclipse) to auto format...  I'm happy to follow anything reasonable :)

In regards to Tests and running PostgreSql.   Yes, only way I've found to run PostgreSql is through docker and yes in Travis you can't run PostgreSql on default port (5432), as Travis already provides PostgreSql running on that port.   I've typically had to run my PostgreSql docker instances on different port to get working in Travis.

Yes, we need some way to provide a "default thread" for the DefaultNioLoop.  I've gone with the easiest approach of creating new thread.  In regards to supporting JEE, will be interesting to see what happens in this space.  JEE is fundamentally thread-per-request (and makes heavy use of ThreadLocal variables) and I don't see this changing anytime too soon.   I'm expecting ADBA would actually be looking to better support Async Servlets or some form of that.   I've not kept too abreast of the JEE space (or I believe Jakarta EE space now).  

But that all aside, there will be need to have a thread available to run "callbacks" (or "then's" of promises, complete stage handlers, etc).   I'd be looking to ADBA specification on how this thread is provided to service "callbacks".   But very happy to include ExecutorService if that is more preferred.

As per exposing the NioLoop, I'm just following JPA in this regards.   You can wire in DataSource implementations to be used with JPA rather than JPA instantiating one for you.   I'm just following that pattern here to inject an NioLoop.   Either way, it needs to be available to the DataSource before the first connection is made.  Swapping NioLoops mid connection mean moving channels to potentially different Selector that could get very tricky :)   Going here for simplest existing pattern.   However, very happy for it to evolve as the use cases start coming forword   (only one I can think of is multiple DataSources sharing the same NioLoop with HTTP servicing for true "one thread" NIO)

I'm looking at pipelining in regards to use case of report generation.  Typically, data needs to be extracted from multiple sources with multiple queries.   By being able to fire off all the requests straight after each other (pipelining) rather than waiting for their responses, it will improve overall latency of the report generation.   However, yes, once transactions and updates get involved that alter the state of the connection and the data...  this becomes a lot more complicated (and pipelining may need to be held up to sequence requests appropriately).

**Comment by sagenschneider** [comment]
This reporting is the use case for what I'm interested in for ADBA (so can pull data from multiple databases of differing vendors by OfficeFloor... with it's much more flexible threading models via thread injection - http://officefloor.net/about.html ).

Also, slightly unashamedly, I'm looking for a high performance PostgreSql driver capable of reaching between 700K-800K selects per second for the TechEmpower benchmark tests.   I'm not finding anything out there, so looking to help build one :)

**Comment by davecramer** [comment]
On 2 August 2018 at 06:09, Daniel <notifications@github.com> wrote:

> This reporting is the use case for what I'm interested in for ADBA (so can
> pull data from multiple databases of differing vendors by OfficeFloor...
> with it's much more flexible threading models via thread injection -
> http://officefloor.net/about.html ).
>
> Also, slightly unashamedly, I'm looking for a high performance PostgreSql
> driver capable of reaching between 700K-800K selects per second for the
> TechEmpower benchmark tests. I'm not finding anything out there, so looking
> to help build one :)
>
So when you want 7-800k selects per second can you give me more
information. I would think this is not terribly difficult to obtain
depending on the selects.

What is the limiting factor here ?



Dave Cramer

**Comment by sagenschneider** [comment]
@davecramer  Hikari explains this reasonably well:  https://github.com/brettwooldridge/HikariCP/wiki/About-Pool-Sizing

Summary of details is optimal connection pool size is the following formula:

    (number of database cores * 2) + number of spindles

Though with SSD number of spindles these days is typically 0.  Note: this formula actually originates from PostgreSql.


In relating this to my understanding of PostgreSql:

- PostgreSql creates a server process per connection
- From a servicing point of view it always faster to do two CPU bound operations one after each other than in parallel (due to thread/process context switching)
- With SSD and in memory caching of data, ideally you want the number of processes to match the number of cores for maximum through put (double if hyperthreading available in our case due to I/O of request/responses)

So it is best from database optimisation point of view to keep connections down. Basically avoids slow down due to lock contentions, waits and thread context switching.


However, from a network point of view, there is round trip latency.  Therefore, to increase network throughput you need to send more requests.  So far with most drivers, the only way to do this is create more connections.


The result is the two opposing forces reduce through put.  Hence, why the Vertx pipeline driver (reactive-pg-client) for PostgreSql gets 150K requests higher than the next competitor.


Pipelining stops the opposing forces:
- connection is fully saturated with requests sent straight after each other (so get full use of band width)
- one connection per CPU means optimal database operation (plus pipelined requests also ensure there is something for the database to process next without waiting on network requests to come through)

There are other factors such as TCP header overheads, nagle algorithm improvements, etc.  However, the above is the key item why I'm looking to ADBA for a fast low latency solution for heavy lifting of data for reporting.

Plus if we can get more than 700K requests per second by a 14 CPU web application server talking over 10Gb network to 14 CPU database server...  then we effectively get top of the TechEmpower benchmarks for 3 out of the 6 benchmarks tested :)  But looking at the top contender implementations there is a reason one stands out 150K faster than the next competitor... and to my understanding is making the best use of the above.

**Comment by sagenschneider** [comment]
@alexanderkjall @davecramer  I'm not overly hearing to many issues with #18 and #19 .   I'll get started on refactoring to the two layer architecture (with Operations implementing NetworkRequests that encapsulate the data for that operation)...  leaving the Network layer to optimise getting the data over the network to PostgreSql and response data back into the Operations).

Note: I'll do this on a separate branch and merge into this PR when done to avoid the noise.   Unfortunately, it may some merge fun (but I'm to be done quick enough that too much an issue)

**Comment by alexanderkjall** [comment]
Sounds good. One thing to keep in mind is that it's not possible to know how the network request will looks like when the Operation is created, as that depends on the state of the prepared statement cache.

To make it more clear, imagine the following scenario:

``
    String sql = "select 1 as t";
    try (Connection conn = ds.getConnection()) {
      conn.rowOperation(sql).submit();
      conn.rowOperation(sql).submit();
    }
``

The first operation will do a full Parse-Bind-Execute sequence on the wire, but the next one will detect that we already have a prepared statement in the cache, and will only do the Bind-Execute steps.

This is a good resource to get up to speed on how the postgresql wire format works: https://www.pgcon.org/2014/schedule/attachments/330_postgres-for-the-wire.pdf

**Comment by sagenschneider** [comment]
Thanks, very useful.  To me this sounds a great fit for asynchronous functions:

```Java
    ParseCommand parse = preparedQueryCache.get(sql);
    if (parse == null) {
        parse = new ParseCommand(sql);
        preparedQueryCache.cache(parse);
    }
    parse.then(new BindCommand(... params).then(new ExecuteCommand()).then(new SyncCommand()).send();
```

With this I might actually suggest actually three layers:

1. ADBA Operation layer that provides the api to chain operations together to ensure appropriate sequence of execution
2. Command layer that translates Operations into Commands to be executed (this maintains state of connection and will know if ParseCommand already sent - or, for example, if unknown whether to be sent again) 
3. Network layer to get the Commands onto the wire and returned frames back up the layers via callbacks

And side note, extend in the PDF indicates it is setup for pipelining the commands 😊

**Comment by davecramer** [comment]
For what it's worth I ran the benchmark https://www.google.com/url?q=https://github.com/reactiverse/reactive-pg-client/blob/master/src/benchmark/java/io/reactiverse/pgclient/RawBenchmark.java&sa=D&source=hangouts&ust=1533386636816000&usg=AFQjCNHLbN4vSuo5asLFbWU1yFP-AyQatg

and the pgjdbc driver is actually more performant.

**Comment by sagenschneider** [comment]
@davecramer  Given the nature of the test I would very much expect pgjdbc driver to be much faster.

Note: I hope I'm not coming across saying JDBC drivers will be obsoleted by ADBA or they are inferior in any means.   They are just different purposes.

Firstly my take on the test you linked, is it would be running to ensure the asynchronous driver is trying to get close to JDBC driver, but I would not expect it ever to beat for following reasons:
- network is localhost so has effectively negligible latency, hence in this test always saturated
- as testing a single connection on a single thread, it sets up well for serialised queries to the database optimising database performance
- the result, it comes down to which driver has less machine operations to get through.
- the JDBC driver benefits in the efficiency of the O/S handling threading via blocking sockets...   where as, the asynchronous driver needs to co-ordinate thread synchronising to get the results back to the main thread..   i.e. the latch.get(..)
- JDBC also benefits from being procedural and little object creation (most memory on thread stack), while the asynchronous driver has to create callback objects (more memory in the heap)...  hence there is going to be more GC pressure for the asynchronous driver
- so I certainly would be running this test...   but it's purpose, as I see it, is not to beat the JDBC driver but rather find instruction inefficiencies to reduce machine operations (and object creation) of the asynchronous driver to get as close as possible to JDBC


Now my thoughts on the usefulness of JDBC and ADBA (as I actually want both, just ADBA is certainly not as mature and is answering a growing trend of functional programming that I find is the pendulum swinging into something new and shiny a little too far until it settles back down - and actually nothing overly new as functional programming has been around for a long time... try some lambda calculus and consider sequencing functions via monads ;) ... so nothing new, just more mainstream available... but I'm getting off topic, so back to JDBC vs ADBA):
- applications don't run database queries in isolation.  JPA follows the graph of the object that requires the identifiers from the previous query to execute the next query.  ADBA will not really help in this type of application, as there is natural dependencies between queries and the above efficiencies will likely have JDBC out perform ADBA on this
- caveat: if the graph has many branches that can be traversed in parallel then ADBA can do this in parallel.  I, however, would expect for majority of applications the ease of development with JPA on JDBC would out way the few times the application could benefit from this.  Especially as efficiencies are negligible (it would only save milliseconds of time and human reaction time is around a quarter of a second)
- and on that caveat, this is where ADBA can be very useful.   Reporting, as I've already mentioned, would benefit from asynchronous extract of data via many queries co-ordinated (rather than single thread sequentially collecting the data)
- Extract Transform Load (ETL) tools would also I see benefit from ADBA.   Rather than having to open lots of connections to try to get the throughput to load data to a database outside transactions.   They can now run in transactions and "pipe" asynchronous inserts to the database increasing throughput of the loads.
- plus other cases I'm sure, but not to replace JDBC in its role with JPA :)

In summary (as I see it), unless Java is going to introduce "async/await" operations to ease asynchronous development, JPA will continue to be developer choice (and with that JDBC).  And even if Java does introduce it, async/await is messy compared to the ease of JPA with blocking JDBC.  There, however, will be specific isolated use cases for improved efficiencies (that will likely be written by only more senior developers) and be embedded in libraries or tools keeping it away from "mainstream" development (your typical web/REST application on a relational database).

And one case of a speciality library, is a report service I want to build for OfficeFloor.  You configure a graph of queries to be executed with the dependencies on each other, and the engine asynchronously gathers all the data for you.  It will use either ADBA or JDBC based on availability of drivers for the various databases storing the report data - especially with micro-service trends.

And some background on where I'm coming from:

OfficeFloor has a lot more flexibility with it's threading model (via continuation injection and thread injection) that it allows mixing together blocking and non-blocking I/O.  You are not forced into just one paradigm of thread-per-request or asynchronous that the various application/web servers enforce on you.

As OfficeFloor blends both models natively, you get choice to solve the problems with the appropriate paradigms.  Rather than this being enforced on you by your web/application server, and this choice forcing ADBA to become a "need", you get choice in OfficeFloor.  In the above engine, ADBA would be preferred over JDBC, however, OfficeFloor can fall back to JDBC if required.  But for managing the data by each micro-service I'd certainly recommend JPA on JDBC.  And this could all run seamlessly in the same OfficeFloor application.

Note: to achieve supporting a blended model, read the paper ( http://www.officefloor.net/about.html - free download ).   The paper actually comes to the conclusion that the method signature is too coupled forcing this choice.   OfficeFloor changes this and from this gives you the freedom of choosing different threading models in the same application (and even to service different parts of the same request).

I'm not for asynchronous or synchronous programming.  I'm actually for complimenting Dependency Injection with Continuation Injection and Thread Injection for a more complete Inversion of Control...  so that asynchronous/synchronous programming is not a fundamental architecture choice, but rather just a tool to solve particular problems within your application.  Hence, just as functional programming is now really moving out of academia to mainstream, my interested is to move Continuation Injection and Thread Injection out of academia to be more mainstream.

Ok, shameless advertising over and I'll stop before writing a "war and peace" novel as certainly can keep going, but I hope this helps you see where my interests are :-)

So yes, pgsql with JPA will outperform ADBA for typical web/REST application use.  ADBA, in these cases, only provides negligible performance improvements that are thwarted by the more complicated programming model (an inaccurate application is worse than a negligibly slower application).  However, there are use cases for ADBA in more isolated problems that I'm interested in pursuing this driver :)  pgsql I believe is mature and well in hand :)

**Comment by davecramer** [comment]
@sagenschneider Thanks for the explanation. I had come to the realization that async gets faster once the latency of channel increases and if you can saturate the channel.

At this point I'm somewhat interested in figuring out how to get the same with pgjdbc ...

**Comment by sagenschneider** [comment]
:)   I bring you back to:

```Java
    Connection conn = ...;
    for (int i = 0; i < 100; i++) {
        new Thread(() -> conn.createStatement("INSERT ...").executeUpdate()).start();
    }
```

If this could pipeline requests over the connection to saturate it, then you can get the improvement.

However, unfortunately the nature of the JDBC API forces blocking:

```Java
    ResultSet resultSet = statement.executeQuery();
```

The thread needs to block to be able to return a synchronous result.  To become asynchronous, then a callback needs to be providing and no synchronous return:

```Java
    statement.executeQuery((error, results) -> { ... });
```

This allows the invoking thread to make multiple calls without blocking.

So if pgsql could not synchronize and allow multiple statements for different threads without blocking, then there is potential for saturating the connection.

OfficeFloor has uses for this:
- read-only connection that is shared amongst all requests for reading in data ( https://github.com/sagenschneider/OfficeFloor/blob/b05936a25382ff9d242e59f02fc135dd21465726/officefloor/persistence/officejdbc/src/main/java/net/officefloor/jdbc/ReadOnlyConnectionManagedObjectSource.java#L41 )
- Wrapping DataSource that re-uses the connection across JPA instances until the EntityManager starts a transaction and then locks a connection to the EntityManager for the duration of the transaction.  This speeds up read operations, which in most applications occurs much more than writes ( https://github.com/sagenschneider/OfficeFloor/blob/b05936a25382ff9d242e59f02fc135dd21465726/officefloor/persistence/officejdbc/src/main/java/net/officefloor/jdbc/pool/ThreadLocalJdbcConnectionPool.java#L414 )

**Comment by alexanderkjall** [comment]
@sagenschneider I have just pushed some initial support for TLS to the driver, and I fear that this will conflict a bit with your work. The TLS code is still a bit of a mess, and I'm planning to go over it and improve it a lot more, but I felt that it's best to not let perfection be the enemy of progress and pushed it.

If your rewrite is severly blocked by this I don't see a problem with rewriting the TLS support one more time, this was the first time I used the SSLEngine class.

**Comment by sagenschneider** [comment]
@alexanderkjall   I've split out ProtocolV3 into NetworkRequest / NetworkResponse objects so that state is managed more as a pipeline (i.e. state can change with multiple in flight queries pipelined over the connection).  To get the performance I'm looking for and be able to provide async nio running multiple queries, I'm thinking there is not one state for the connection (for pgjdbc this may be the case due to blocking for response....  but for pgsql2...  there could be multiple requests in flight with multiple responses being waited on...  bit like TCP windows).

I'm currently in the process of hooking it up to run in the officefloor-raw tests within the TechEmpower benchmark tests.  This can get us early feedback that the network layer architecture is performant (and avoid later possible significant restructuring if it is not).

Work still to do on this branch:

1. Some tests are still failing.  These are mostly group operations and some error handling.  This is above the network layer and I could do with your help to get them working.

1. The BEFrame returns an byte[] for getPayload().   I want to change this to InputStream getPayload() with possible additional helper method int getPayloadSize().   This will allow streaming the data directly out of ByteBuffer instances to avoid unnecessary copying into arrays.

Note: once BEFrame is providing InputStream for getPayload(), it is much easier to implement TLS as this can be under the hood at the network layer.  It follows:

Without TLS:

NetworkRequest -> NetworkOutputStream -> ByteBuffer -> Socket
NetworkResponse <- NetworkInputStream <- ByteBuffer <- Socket

Then with TLS, it is hidden from NetworkRequest / NetworkResponses:

NetworkRequest -> NetworkOutputStream -> ByteBuffer -> SSL Context / Engine -> ByteBuffer -> Socket
NetworkResponse -> NetworkInputStream <- ByteBuffer <- SSL Context / Engine <- ByteBuffer <- Socket

**Comment by alexanderkjall** [comment]
I'll have a look at fixing the failing unit tests.

Regarding TLS, the SSLEngine, I don't really follow why it should be easier after changing to an InputStream, but maybe it will become apparent once I try to write the code :)

About state of the connection, after thinking a bit about it I realize that it's mostly the handshake and login process, what the state is when queries are in flight isn't as complicated.

**Comment by sagenschneider** [comment]
@alexanderkjall   I'm just in the process of getting the InputStream working to avoid undue object creation (creates extra GC pressure, slowing performance).

I've also hooked pgsql2 up to performance tests.  I, however, am finding it fails after significant load (typically queries size is 500).  The code is not yet merged into the main tests, as need to get passing.  My guess is it is my code, however, would be good to get running in this environment to ensure any threading / stress bugs are identified and fixed  (plus this runs continuously so good performance tracking as we change things):   https://github.com/sagenschneider/FrameworkBenchmarks/blob/df71a9fd83d4db9e28c08e93847992df1e7aba29/frameworks/Java/officefloor/src/woof_raw/src/main/java/net/officefloor/benchmark/RawOfficeFloorMain.java#L298

Again, I'm still catching up on the ADBA specification.   So if I'm using wrong please let me know.

**Comment by alexanderkjall** [comment]
@sagenschneider The failure sounds troubling, one thing that I haven't implemented yet is heuristics for retiring portals on the server side. There needs to be some heuristics regarding how they are managed, maybe a LRU-cache.

I haven't had much free time recently, but I have fixed some more of the failing unit tests and are currently debugging the test tryToAddRowPublisherOperationAfterProhibit. There is some problem with the logic around the class ImmediateComplete, it makes the request complete to quick, should complete "at it's turn" when the bytes are read, and instead it's completed at write time. I haven't solved how to fix it yet, do you have any suggestions?

**Comment by sagenschneider** [comment]
@alexanderkjall   No worries.   I'm caught up with the InputSteam changes (and not getting the time I'd like to devote to it too at the moment either).

Thanks for looking into the tests.   Must excuse me, but I'll look at the tests once I've done the InputStream work.   This way only need to tidy up the tests once.  If tests are taking a bit of an effort to fix, my suggestion is leave them until I've done the InputStream and we can then work through each to get working (as network layer should change little after this point - I hope).

The network layer has means to co-ordinate requests/responses:
  - request can be written and flag blocking so no further request is written to the network until the provided response is fulfilled (useful for connection setup... likely little use for pipeline... unless possible close conversation is required)
  - there is possibility to send another request immediately after an existing request.   This is typically so the parse, bind, execute requests follow each other (and don't interlace other requests).  Possibly something not necessary but allows to get all requests out onto the wire in order.  Each request then registers an expected response in the order the requests were made.
  - responses can trigger further writes (again more for connection and close handshakes)

I'm not sure of the "ImmediateComplete" class.   I'll have to have a look once I've got the InputStream fixes in.   However, read/writes can happen very close to each other.  I would not be thinking in state change of read then write.   Due to selector being event triggered, the read/writes can happen in any random picked order (i.e. if a write is registered it could be written before the next read occurs).  Best to think of the output and input streams as two separate things.   Sorry, can't be of more help at moment.  When I get some time I'll look into the issue.

However, for now (given tests are taking some effort to fix), I would hold off on fixing them until I get the InputStream work in.   Then happy to work with you to get all the tests working and this PR merged :)

**Comment by sagenschneider** [comment]
@alexanderkjall   and yes, sounds like we might need to close the portals once the request is complete.   Note sure heuristics is the helper here in regards to an LRU cache.   I would be expecting the portal be complete once all the data has been retrieved (essentially a close on the operation chain).

My thinking is that there is some "implicit" close of the portal once the chain has completed execution.  However, I'm not too clear on what portal really means, so happy to be corrected.

**Comment by sagenschneider** [comment]
@davecramer  how does pgjdbc handle portals?

**Comment by alexanderkjall** [comment]
@sagenschneider A portal can be closed directly after it's used, but that would be wasteful, a portal is the server side of what's also called a prepared statement, so by reusing them we save the time it takes to parse the sql string on the server.

**Comment by davecramer** [comment]
@sagenschneider that's a fairly open ended question. Can you be a bit more specific ?

as @alexanderkjall eluded to they are the server side answer to a prepared statement. 

most of the time we use the un-named portal until a statement has been used 5 times then we name the portal and keep it around.

It is also used for implementing fetch count

**Comment by sagenschneider** [comment]
@alexanderkjall @davecramer   Oh, my bad.   Think I've got query name and portal name mixed around.

Is it, therefore, as follows?
- portal name identifies the prepared statement for re-use (to avoid parse stage again)
- query name identifies the instance of the query (so that further fetches of data can identified)

**Comment by davecramer** [comment]
@sagenschneider 

I'm not sure what a query name is. 
A portal however is what you get back after you do a prepare on the server side. Although confusing matters it is also what you get when you create a cursor. That said both a cursor and a prepared statement which return results are basically the same thing so returning a portal makes some sense here

**Comment by alexanderkjall** [comment]
@sagenschneider closing this one as I merged the commits locally, great work.

**Pull request closed without merging**